### DISCUSSION TITLE
[java] @ThisModifier compare not working in XPath

### DISCUSSION THREAD

**Discussion by d1ss0nanz** [pmd/pmd#3045] [Q&A] [answered]
**Affects PMD Version:** 
6.0.2

**Description:**
I'm trying to write a custom XPath rule that is based on detecting
`//PrimaryExpression[PrimaryPrefix/@ThisModifier='true']
`
Doesn't match. Neither does:
`//PrimaryExpression/PrimaryPrefix[@ThisModifier='true']
`
While in designer `//PrimaryExpression/PrimaryPrefix `selects the right nodes and _"ThisModifier = true"_.

**Code Sample demonstrating the issue:**

```
public class MockCOD_Enum extends MockCOD
{
    private EnumTestMode testEnum = EnumTestMode.ENUM_1;
    private List<EnumTestMode> enumList = new ArrayList<>();
    private List<EnumTestMode> enumListNonEmpty =
        new ArrayList<>(Arrays.asList(EnumTestMode.ENUM_1, EnumTestMode.ENUM_2));

    @Override
    public int hashCode()
    {
        final int prime = 31;
        int result = super.hashCode();
        result = prime * result + ((testEnum == null) ? 0 : testEnum.hashCode());
        return result;
    }

    @Override
    public boolean equals(final Object obj)
    {
        if (this == obj)
        {
            return true;
        }
        if (!super.equals(obj))
        {
            return false;
        }
        if (!(obj instanceof MockCOD_Enum))
        {
            return false;
        }

        final MockCOD_Enum other = (MockCOD_Enum) obj;
        return testEnum == other.testEnum;
    }

    /**
     * TODO: JavaDoc
     * 
     * @return TODO
     */
    public EnumTestMode getEnum()
    {
        return this.testEnum;
    }

    /**
     * TODO: JavaDoc
     * 
     * @param testMode TODO
     */
    public void setEnum(final EnumTestMode testMode)
    {
        testEnum = testMode;
    }
}

```

**Running PMD through:** 
Designer

**Comment by oowekyala** [comment, marked as answer]
If you're using XPath 2.0 (the default in the designer), then you should know that the boolean constants are written `true()` and `false()`. Only XPath 1.0 uses `'true'` and `'false'`, which are string literals in XPath 2.0. I wonder why this doesn't throw a type error though.

**Comment by d1ss0nanz** [comment]
Ok, that helped.

This works:
`//PrimaryExpression/PrimaryPrefix[.[@ThisModifier=true()]]`
